jit/core/jit_cleaner.py

```diff
@@ -0,0 +1,160 @@
+import json
+from datetime import datetime
+
+import google.auth
+import googleapiclient.discovery
+from google.api_core import retry
+from google.cloud import pubsub_v1
+from google.oauth2 import service_account  # type: ignore
+
+from jit.utils import config
+from jit.utils import constant
+from jit.utils.logger import jit_logger
+
+
+def modify_policy_remove_member(conf: config.JitConfig, target_project: str,
+  role: str, member: str,
+  start: datetime, end: datetime):
+    """Removes a  member from a role binding."""
+    jit_logger.info(
+      "modify_policy_remove_member. config=%s, target_project=%s, role=%s, member=%s, start=%s, end=%s",
+      conf, target_project, role, member, start, end
+    )
+    credentials, _ = google.auth.default()
+    service = googleapiclient.discovery.build(
+      "cloudresourcemanager", "v1", credentials=credentials
+    )
+    policy = (
+      service.projects()
+      .getIamPolicy(
+        resource=target_project,
+        body={"options": {"requestedPolicyVersion": 3}},
+      )
+      .execute()
+    )
+    jit_logger.debug(
+      "retrived policy. policy=%s, project=%s", policy, target_project
+    )
+    expression = f"(request.time >= timestamp(\"{start}\") && request.time < timestamp(\"{end}\"))"
+    if len(policy.get("bindings", [])) == 0:
+        jit_logger.error("project binding empty. ")
+        return
+    binding_count = 0
+    for b in policy["bindings"]:
+        if (
+          role == b["role"] and
+          member in b["members"] and
+          expression == b.get("condition", {}).get("expression", "")
+        ):
+            b["members"].remove(member)
+            binding_count += 1
+            continue
+
+    policy["bindings"][:] = [b for b in policy["bindings"] if b["members"]]
+    # if "members" in binding and member in binding["members"]:
+    #     binding["members"].remove(member)
+    # jit_logger.info(binding)
+
+    service.projects().setIamPolicy(
+      resource=target_project,
+      body={"policy": policy}
+    ).execute()
+    jit_logger.info("%d binding has been removed.", binding_count)
+
+
+def process_pubsub_msg(conf, received_message):
+    jit_logger.info(f"Received: {received_message.message.data}.")
+    jit_logger.info(f"attributes: {received_message.message.attributes}")
+    origin = received_message.message.attributes.get("origin", "")
+    # for invalid origin, we should directly return and make the run_jit_cleaner
+    # ack the message.
+    if not origin or constant.MessageOrigin.from_str(
+      origin) != constant.MessageOrigin.BINDING:
```

Comment:
hmm, we were meant to filter out messages based on the subscription. This is a good catchall but it's no aligned to the high level design.

Reply:
same issues with this one. looks like failed to get your point, may need your further explain

Reply:
updated the comment